src/eventbridge/eventpattern/index.ts

```diff
@@ -0,0 +1,798 @@
+  Expr,
+  PropAccessExpr,
+  UnaryExpr,
+} from "../../expression";
+import {
+  assertDefined,
+  assertNever,
+  assertNumber,
+  assertString,
+} from "../../assert";
+import { isReturn, Stmt } from "../../statement";
+import {
+  PatternDocument,
+  Pattern,
+  isPatternDocument,
+  isNumericAggregationPattern,
+  isNumericRangePattern,
+  isPresentPattern,
+  isAggregatePattern,
+  isAnythingButPattern,
+  isExactMatchPattern,
+  isPrefixMatchPattern,
+  isEmptyPattern,
+  patternDocumentToEventPattern,
+} from "./pattern";
+import * as fnls_event_bridge from "./types";
+import {
+  joinNumericRange,
+  reduceNumericAggregate,
+  mergeNumericOr,
+  negateNumericRange,
+  andNumericAggregation,
+  andNumericAggregationWithRange,
+  createSingleNumericRange,
+} from "./numeric";
+import {
+  BinaryOp,
+  isBinaryExpr,
+  isBooleanLiteral,
+  isCallExpr,
+  isElementAccessExpr,
+  isIdentifier,
+  isNullLiteralExpr,
+  isNumberLiteralExpr,
+  isPropAccessExpr,
+  isStringLiteralExpr,
+  isUnaryExpr,
+} from "../..";
+import {
+  INCLUDES_SEARCH_ELEMENT,
+  OPERATIONS,
+  STARTS_WITH_SEARCH_STRING,
+} from "./constants";
+import { invertBinaryOperator } from "./utils";
+
+/**
+ * https://github.com/sam-goodwin/functionless/issues/37#issuecomment-1066313146
+ */
+export const synthesizeEventPattern = (
+  predicate: FunctionDecl
+): fnls_event_bridge.FnLsEventPattern => {
+  const eventDecl = predicate.parameters[0];
+
+  const handleExpr = (expr: Expr): PatternDocument => {
+    if (isBinaryExpr(expr)) {
+      return handleBinary(expr);
+    } else if (isPropAccessExpr(expr)) {
+      return handlePropAccess(expr);
+    } else if (isUnaryExpr(expr)) {
+      return handleUnaryExpression(expr);
+    } else if (isCallExpr(expr)) {
+      return handleCall(expr);
+    } else {
+      throw new Error(`${expr.kind} is unsupported`);
+    }
+    // assertNever(expr);
+  };
+
+  const handleSmnt = (stmt: Stmt): PatternDocument => {
+    if (isReturn(stmt)) {
+      return handleExpr(stmt.expr);
+    } else {
+      throw new Error(`Unsupported statement ${stmt.kind}`);
+    }
+  };
+
+  const handleBinary = (expr: BinaryExpr): PatternDocument => {
+    if (expr.op === "==") {
+      return handleEquals(expr);
+    } else if (expr.op === "!=") {
+      return handleNotEquals(expr);
+    } else if ([">", ">=", "<", "<="].includes(expr.op)) {
+      return handleNumericRange(expr);
+    } else if (expr.op === "in") {
+      return handleInOperation(expr);
+    } else if (expr.op === "&&") {
+      return handleAnd(expr);
+    } else if (expr.op === "||") {
+      return handleOr(expr);
+    } else {
+      throw new Error(`Unsupported binary operator ${expr.op}`);
+    }
+  };
+
+  const handleAnd = (expr: BinaryExpr): PatternDocument => {
+    const left = handleExpr(expr.left);
+    const right = handleExpr(expr.right);
+
+    return andDocuments(left, right);
+  };
+
+  const andDocuments = (
+    classDocument1: PatternDocument,
```

Comment:
what does `classDocument` mean?

Reply:
Had been calling `Patterns` "Classifications", but have changed them to `Patterns` everywhere now.